src/sign/ring.rs

```diff
@@ -207,6 +213,121 @@ impl<'a> Sign for SecretKey<'a> {
     }
 }
 
+//------------ Nsec3HashError -------------------------------------------------
+
+/// An error when creating an NSEC3 hash.
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub enum Nsec3HashError {
+    /// The requested algorithm for NSEC3 hashing is not supported.
+    UnsupportedAlgorithm,
+
+    /// Data could not be appended to a buffer.
+    ///
+    /// This could indicate an out of memory condition.
+    AppendError,
+
+    /// The hashing process produced an invalid owner hash.
+    ///
+    /// See: [OwnerHashError](crate::rdata::nsec3::OwnerHashError)
+    OwnerHashError,
+}
+
+/// Compute an [RFC 5155] NSEC3 hash using default settings.
+///
+/// See: [Nsec3param::default].
+///
+/// [RFC 5155]: https://www.rfc-editor.org/rfc/rfc5155
+pub fn nsec3_default_hash<N, HashOcts>(
+    owner: N,
+) -> Result<OwnerHash<HashOcts>, Nsec3HashError>
+where
+    N: ToName,
+    HashOcts: AsRef<[u8]> + EmptyBuilder + OctetsBuilder + Truncate,
+    for<'a> HashOcts: From<&'a [u8]>,
+{
+    let params = Nsec3param::<HashOcts>::default();
+    nsec3_hash(
+        owner,
+        params.hash_algorithm(),
+        params.iterations(),
+        params.salt(),
+    )
+}
+
+/// Compute an [RFC 5155] NSEC3 hash.
+///
+/// Computes an NSEC3 hash according to [RFC 5155] section 5:
+///
+/// > IH(salt, x, 0) = H(x || salt)
+/// > IH(salt, x, k) = H(IH(salt, x, k-1) || salt), if k > 0
+///
+/// Then the calculated hash of an owner name is:
+///
+/// > IH(salt, owner name, iterations),
+///
+/// Note that the `iterations` parameter is the number of _additional_
+/// iterations as defined in [RFC 5155] section 3.1.3.
+///
+/// [RFC 5155]: https://www.rfc-editor.org/rfc/rfc5155
+pub fn nsec3_hash<N, SaltOcts, HashOcts>(
+    owner: N,
+    algorithm: Nsec3HashAlg,
+    iterations: u16,
+    salt: &Nsec3Salt<SaltOcts>,
+) -> Result<OwnerHash<HashOcts>, Nsec3HashError>
```

Comment:
> Why not just take the owner name and an `Nsec3param`?

I had that but took it out. There was a reason. I'll see if I can remember why.

Reply:
Ah I know why. The `nsec3s()` function takes an `Nsec3Param` struct which it uses to create the `NSEC3PARAM` RR at the apex of the zone. For the `NSEC3PARAM` RR RFC 5155 says that the opt-out flag in the flags field MUST be zero, so for RFC compliance the `Nsec3Param` struct passed to `nsec3s()` should have the opt-out flag set to zero.

Honestly I think this is a bit of a foot-gun and perhaps best not to pass an `Nsec3Param` to `nsec3s()` but instead only the other fields (algorithm, iterations, salt), however MAYBE in future it will be legal to set some of the other flag bits in the flags field and a user would want to have those set in the created `NSEC3PARAM` RR... so for that reason `nsec3s()` currently takes an `Nsec3Param` as input.

When generating NSEC3 RRs and when opt-out is enabled the flags value in the given `Nsec3Param` cannot be used as-is because the opt-out flag must be set to 1 (but NOT in the `NSEC3PARAM` RR), ad rather than copy the given `Nsec3Param` or modify it and then pass it to `nsec3_hash()` I felt it was better to just pass only the values actually needed for hashing in, as NSEC3 hashing doesn't need the flags field at all, and also that way users don't have to think about what value to set the unused `Nsec3Param::flags` field to when invoking `nsec3_hash()` directly (as `dnst nsec3-hash` does).